lmdeploy/pytorch/engine/engine.py

```diff
@@ -1003,9 +1003,29 @@ def __send_resps(step_outputs: Dict[int, InferOutput]):
             for out in step_outputs.values():
                 __send_resp(out)
 
+        def __do_prefill():
+            # decoding if no waiting
+            if not self.scheduler.has_waiting():
+                return False
+
+            num_running = len(self.scheduler.running)
+            num_waiting = len(self.scheduler.waiting)
+            max_batches = self.scheduler_config.max_batches
+
+            # prefill if too much waiting
+            if num_waiting >= 4:
+                return True
+
+            # prefill if no enough running
+            if num_running < max_batches * 0.5:
+                return True
+
+            # decoding
+            return False
+
         async def __step():
             """step decoding."""
-            prefill = self.scheduler.has_waiting()
+            prefill = __do_prefill()
```

Comment:
What's the benefit of this change?

Reply:
This is added by for better schedule policy @grimoire

Reply:
prefill will not enable cudagraph, short prefill would lead to long kernel launch overhead. So we will not do prefill unless we have enough batch size.